### DISCUSSION TITLE
getValues() doesn't represent provided form generic

### DISCUSSION THREAD

**Discussion by tykhan** [react-hook-form/react-hook-form#11724] [Q&A] [answered]
Hey!
I have next question. Assume, I provided a generic to useForm hook, but I'm not forced to provide defaultValues. In that case, my initial form would be an empty object and any attempt to interact with it will lead to uncaught errors. As you see on screenshot typescript says 'foo' = string, but in reality it's undefined. 
Am I doing something wrong or it is intended?

<img width="567" alt="image" src="https://github.com/react-hook-form/react-hook-form/assets/62352586/7d3ae5e4-be15-4387-9169-909d98f05fed">
<img width="528" alt="image" src="https://github.com/react-hook-form/react-hook-form/assets/62352586/d1ef202f-a71d-481f-a4da-75a8707c4cae">

Also, if I want to provide default values, they are optional even I provided a generic where they are required
<img width="524" alt="image" src="https://github.com/react-hook-form/react-hook-form/assets/62352586/f0b8893f-883c-4acb-986c-785095d4427c">

**Comment by sujinleeme** [comment]
Which react hook form do you use now? 

The defaultValue type uses DeepPartial, which means all properties of an existing type(form value) are optional.
https://github.com/react-hook-form/react-hook-form/blob/306627e385738e322cad8dff9e752678f8210020/src/types/form.ts#L47

The latest version(v7)  has no problem with defining the FormData type
<img src="https://github.com/react-hook-form/react-hook-form/assets/6011401/2b3a0616-caf8-48ac-8a0d-244e7f699899" width="50%" />

```ts
  const { getValues } = useForm<{
    foo: string;
  }>({
    mode: "all",
    defaultValues: {
      foo: undefined,
    },
  });

  // You may need to use the optional chaining operator because the value can be undefined.
  const fooLength = getValues().foo?.length; 
  ```

**Reply by tykhan** [reply]
@sujinleeme , thank for reply! I'm on the latest version

> // You may need to use the optional chaining operator because the value can be undefined.
>   const fooLength = getValues().foo?.length; 

Well, the problem here that typescript allows you NOT to use optional chaining, since it thinks it can be only string. And for sure it may lead to uncaught errors.

The question is actually, why defaultValues uses DeepPartial?
If I'd like my 'foo' to be undefined I would create appropriate type like
`type MyFormT = { foo: string | undefined }`

**Reply by sujinleeme** [reply]
The reason for using DeepPartial is to provide flexibility when initializing the form with default values. 
It enables us to achieve partial initialization of form fields. For instance, 'foo' can be initialized as undefined, and 'bar' can be entirely omitted. 

```ts
  const { getValues } = useForm<{
    foo: string;
    bar: string
  }>({
    mode: "all",
    defaultValues: {
      foo: undefined, 
    },
  });
```

If we specify default values, they must be provided with deep equality, as in `{ defaultValues: { foo: '' } }`. Or all properties inside `FormValue(MyFormT)` will be optional. It will be hard to keep type safety when dealing with form errors or the onSubmit handler.

<img src="https://github.com/react-hook-form/react-hook-form/assets/6011401/3daa081e-4493-4aec-8ec0-6a1b1a41e50c" width="50%" />

The generic type definition is not related to triggering error.

**Reply by tykhan** [reply]
@sujinleeme  okay, let me rephrase that a bit
Is it expected that foo is typed as string here, even if in reality it is undefined? Don't you find it at least unsafe?

<img width="390" alt="image" src="https://github.com/react-hook-form/react-hook-form/assets/62352586/e5a1f74a-d63a-40f9-982d-d726e2ac5095">

**Reply by sujinleeme** [reply, marked as answer]
Yes, it is expected because you provide the generic. https://github.com/react-hook-form/react-hook-form/blob/ebc1f2c3ee840dc86990483dbf27fb1c43264ed9/src/types/form.ts#L837

If you remove the generic type, react hook form will infer the type based on `defaultValues`

```ts
  const { getValues } = useForm({
    defaultValues: {
      foo: undefined,
    },
  });

  const values = getValues(); 
  ```
  
<img src="https://github.com/react-hook-form/react-hook-form/assets/6011401/10be672c-cde2-44b8-b496-e8f6abad5cd1" width="50%" />

In my experience, the best way to solve this type of mismatching with the generic is just providing  `{foo: ''}` if the field is a simple input component. Then you can avoid this.
